From a697b908a3cfd3f294d5df00bdb301794409758e Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Thu, 28 Apr 2022 20:20:51 +0300
Subject: [PATCH] [SW-0] rom: update rma mode logic

As per the security specification, in rma mode, the pre-production
efuse bits are to be reset to pre-production values . And in any case
key hash index 4 should never be revoked . This patch updates the
logic as per this requirement.

Change-Id: Ie34607fa59683a49b72b00f22cf4ae84c9910540
---

diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 4bb0357..191af2c 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -382,7 +382,11 @@
 	rma_mode_status >>= BOOT_STRAP_RMA_MODE_MASK;
 	/* TODO: add a relavent gpio check for RMA mode once finalized for gaudi3 */
 	if ((rma_mode_status & BOOT_STRAP_RMA_MODE_MASK) == BOOT_STRAP_RMA_MODE_MASK)
-		arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_ENABLE_VALIDATION_MASK;
+		/* Reset the efuse pre-production bits to 0(pre-production value) */
+		arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_PRE_PRODUCTION_BITS;
+
+	/* KEY index 4 must not be revoked */
+	arcboot_rom_global.config_fuse &= ~EFUSE_SYSCFG_REVOKE_KEY_4;
 }
 
 __weak int arcboot_rom_fuse_single_die_config(void)
diff --git a/arch/gaudi3/arc/dev_efuse.h b/arch/gaudi3/arc/dev_efuse.h
index 1379e44..f96dbf0 100644
--- a/arch/gaudi3/arc/dev_efuse.h
+++ b/arch/gaudi3/arc/dev_efuse.h
@@ -12,7 +12,7 @@
 #define EFUSE_SECURITY_FW_VERSION_OFFSET		0x3C
 
 #define EFUSE_SYSCFG_REVOKE_KEY_0			BIT(0)
-#define EFUSE_SYSCFG_PERMIT_KEY_4			~BIT(4)
+#define EFUSE_SYSCFG_REVOKE_KEY_4			BIT(4)
 #define EFUSE_SYSCFG_EOM_MASK				BIT(5)
 #define EFUSE_SYSCFG_TPM_CODE_LOCATION_MASK		BIT(6)
 #define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK	BIT(7)
@@ -47,7 +47,6 @@
 							| EFUSE_SYSCFG_ENABLE_VALIDATION_MASK \
 							| EFUSE_SYSCFG_DISABLE_JTAG_IF_MASK \
 							| EFUSE_SYSCFG_REVOKE_KEY_0 \
-							| EFUSE_SYSCFG_SECURITY_ENABLE_MASK) \
-							& EFUSE_SYSCFG_PERMIT_KEY_4)
+							| EFUSE_SYSCFG_SECURITY_ENABLE_MASK))
 
 #endif /* _GAUDI3_DEV_EFUSE_H_ */
